transport/pool.go

```diff
@@ -0,0 +1,201 @@
+			return conn
+		}
+	}
+	return nil
+}
+
+func (p *ConnPool) Conn(token Token) *Conn {
+	idx := p.shardOf(token)
+	if conn := p.loadConn(idx); conn != nil {
+		return conn
+	}
+	return p.RandConn()
+}
+
+func (p *ConnPool) shardOf(token Token) int {
+	shards := uint64(p.nrShards)
+	z := uint64(token.value+math.MinInt64) << p.msbIgnore
+	lo := z & 0xffffffff
+	hi := (z >> 32) & 0xffffffff
+	mul1 := lo * shards
+	mul2 := hi * shards
+	sum := (mul1 >> 32) + mul2
+	return int(sum >> 32)
+}
+
+func (p *ConnPool) storeConn(conn *Conn) {
+	p.conns[conn.Shard()].Store(conn)
+}
+
+func (p *ConnPool) loadConn(shard int) *Conn {
+	conn, _ := p.conns[shard].Load().(*Conn)
+	return conn
+}
+
+func (p *ConnPool) clearConn(shard int) bool {
+	conn, _ := p.conns[shard].Swap((*Conn)(nil)).(*Conn)
+	return conn != nil
+}
+
+func (p *ConnPool) Close() {
+	p.connClosedCh <- _poolCloseShard
+}
+
+// closeAll is called by PoolRefiller.
+func (p *ConnPool) closeAll() {
+	for i := range p.conns {
+		if conn, ok := p.conns[i].Swap((*Conn)(nil)).(*Conn); ok {
+			conn.Close()
+		}
+	}
+}
+
+type PoolRefiller struct {
+	addr   string
+	pool   ConnPool
+	config ConnConfig
+	active int
+}
+
+func (r *PoolRefiller) initConnPool() error {
+	conn, err := OpenConn(r.addr, nil, r.config)
+	if err != nil {
+		return err
+	}
+
+	s, err := conn.Supported()
+	if err != nil {
+		conn.Close()
+		return fmt.Errorf("supported: %w", err)
+	}
+	ss := s.ScyllaSupported()
+
+	r.pool = ConnPool{
+		nrShards:     int(ss.NrShards),
+		msbIgnore:    ss.MsbIgnore,
+		conns:        make([]atomic.Value, int(ss.NrShards)),
+		connClosedCh: make(chan int, int(ss.NrShards)+1),
+	}
+	conn.setOnClose(r.onConnClose)
+	r.pool.storeConn(conn)
+	r.active = 1
+
+	return nil
+}
+
+func (r *PoolRefiller) onConnClose(conn *Conn) {
+	select {
+	case r.pool.connClosedCh <- conn.Shard():
+	default:
+		log.Printf("conn pool: ignoring conn %s close", conn)
+	}
+}
+
+const fillBackoff = time.Second
+
+func (r *PoolRefiller) loop() {
+	r.fill()
+
+	timer := time.NewTicker(fillBackoff)
+	for {
+		select {
+		case <-timer.C:
+			r.fill()
+		case shard := <-r.pool.connClosedCh:
+			if shard == _poolCloseShard {
+				r.pool.closeAll()
+				return
+			}
+			if r.pool.clearConn(shard) {
+				r.active--
+			}
+			r.fill()
+		}
+	}
+}
+
+func (r *PoolRefiller) fill() {
+	if !r.needsFilling() {
+		return
+	}
+
+	si := ShardInfo{
+		NrShards:  uint16(r.pool.nrShards),
+		MsbIgnore: r.pool.msbIgnore,
+	}
+
+	for i := 0; i < r.pool.nrShards; i++ {
+		if r.pool.loadConn(i) != nil {
+			continue
+		}
+
+		si.Shard = uint16(i)
+		conn, err := OpenShardConn(r.addr, si, r.config)
+		if err != nil {
+			log.Printf("failed to open shard conn: %s", err)
+			continue
+		}
+		if conn.Shard() != i {
+			log.Fatalf("opened conn to wrong shard: expected %d got %d", i, conn.Shard())
+		}
+		conn.setOnClose(r.onConnClose)
+		r.pool.storeConn(conn)
+		r.active++
```

Comment:
Correct me if I'm wrong, but shouldn't we check if there already exists connection to this shard and then store it in pool? Wouldn't it cause any trouble to just throw away the pointer of working connection without calling Close() on it?

Reply:
> Correct me if I'm wrong, but shouldn't we check if there already exists connection to this shard and then store it in pool?

We do, note the line 176.

> Wouldn't it cause any trouble to just throw away the pointer of working connection without calling Close() on it?

It happens only when the connection is closed, first closed then removed.
It could also happen when the connection is broken, not yet implemented.
In both cases a dysfunctional connection can be returned.
It would be handled and retried.

Reply:
We do check i-th connection, but what I meant was the following scenario:
1. r.pool.LoadConn(0) == nil 
2. OpenShardConn opens connection to wrong shard  - conn.Shard = 1 (I don't know how likely it is to happen)
3. But we already have working connection to shard 1 
4. r.pool.StoreConn stores newly opened connection and loses pointer to already working connection to shard 1

I'm afraid we are not able to gracefully close such lost connection.

Reply:
log.Fatalf exits the process, it's a FATAL error if your connections goes to a different shard.